sycl/doc/extensions/ExplicitSIMD/dpcpp-explicit-simd.md

```diff
@@ -0,0 +1,745 @@
+functions will always return `1`.
+
+*Functor kernel*
+```cpp
+using AccTy = cl::sycl::accessor<int, 1, cl::sycl::access::mode::read_write,
+                                 cl::sycl::access::target::global_buffer>;
+class Functor {
+public:
+  Functor(int X, AccTy &Acc) : X(X), Acc(Acc) {}
+
+  EXPLICIT_SIMD void operator()() { Acc[0] += X; }
+
+private:
+  int X;
+  AccTy Acc;
+};
+```
+
+*Lambda kernel and function*
+```cpp
+EXPLICIT_SIMD SYCL_EXTERNAL
+void sycl_device_f(cl::sycl::global_ptr<int> ptr, sycl::intel::gpu::simd<float, 8> X) {
+  sycl::intel::gpu::flat_block_write(*ptr.get(), X);
+}
+...
+  Q.submit([&](cl::sycl::handler &Cgh) {
+    auto Acc1 = Buf1.get_access<cl::sycl::access::mode::read>(Cgh);
+    auto Acc2 = Buf2.get_access<cl::sycl::access::mode::read_write>(Cgh);
+
+    Cgh.single_task<class KernelID>([=] () EXPLICIT_SIMD {
+      sycl::intel::gpu::simd<float, 8> Val = sycl::intel::gpu::flat_block_read(Acc1.get_pointer());
+      sycl_device_f(Acc2, Val);
+    });
+  });
+}
+```
+
+## Implementation restrictions
+Current ESP implementation does not support using certain standard SYCL features
+inside explicit SIMD kernels and functions. Most of them will be eventually
+dropped. What's not supported today:
+- Mixing `EXPLICIT_SIMD` kernels with SYCL kernels in a single source
+- Local accessors. Local memory is allocated and accessed via explicit
+device-side API
+- 2D and 3D accessors
+- Constant accessors
+- `sycl::accessor::get_pointer()`. All memory accesses through an accessor are
+done via explicit APIs; e.g. `sycl::intel::gpu::block_store(acc, offset)`
+- Few others (to be documented)
+
+
+## Core Explicit SIMD programming APIs
+
+The DPC++ Explicit SIMD library defines the following classes to enhance the
+expressiveness for explicit SIMD data-parallel programming while enabling
+efficient mapping to SIMD vector operations on Intel GPU architectures.
+
+### SIMD vector class
+
+The `sycl::intel::gpu::simd` class is a vector templated on some element type. The element type must be vectorizable type. The set of vectorizable types is the set of fundamental SYCL arithmetic types (C++ arithmetic types or `half` type) excluding `bool`. The length of the vector is the second template parameter.
+
+Each simd class object is mapped to a consecutive block of general register
+files (GRF).
+
+```cpp
+namespace sycl {
+namespace intel {
+namespace gpu {
+
+// __vector_type, using clang vector type extension.
+template <typename __Ty, int __N> struct __vector_type {
+  static_assert(!std::is_const<__Ty>::value, "const element type not supported");
+  static_assert(__is_vectorizable_v<__Ty>::value, "element type not supported");
+  static_assert(__N > 0, "zero-element vector not supported");
+
+  static constexpr int length = __N;
+  using type = __Ty __attribute__((ext_vector_type(__N)));
+};
+
+template <int __N>
+using __mask_type_t = typename __vector_type<uint16_t, __N>::type;
```

Comment:
The mask should also be a wrapper around the clang-vector type rather than the clang-vector type itself.

The internal storage should be implementation defined. `uint16_t` is a bad choice for some HW. Nor is it how clang-vector types works (using the same size int as the corresponding vector type used for comparison (e.g. long for double and int for float)).

Reply:
I added a TODO